x-pack/plugins/lens/public/indexpattern_datasource/to_expression.ts

```diff
@@ -7,32 +7,29 @@
 import { Ast, ExpressionFunctionAST } from '@kbn/interpreter/common';
 import { IndexPatternColumn } from './indexpattern';
 import { operationDefinitionMap } from './operations';
-import { IndexPattern, IndexPatternPrivateState } from './types';
+import { IndexPattern, IndexPatternPrivateState, IndexPatternLayer } from './types';
 import { OriginalColumn } from './rename_columns';
 import { dateHistogramOperation } from './operations/definitions';
 
-function getExpressionForLayer(
-  indexPattern: IndexPattern,
-  columns: Record<string, IndexPatternColumn>,
-  columnOrder: string[]
-): Ast | null {
+function getExpressionForLayer(layer: IndexPatternLayer, indexPattern: IndexPattern): Ast | null {
+  const { columns, columnOrder } = layer;
+
   if (columnOrder.length === 0) {
     return null;
   }
 
-  function getEsAggsConfig<C extends IndexPatternColumn>(column: C, columnId: string) {
-    return operationDefinitionMap[column.operationType].toEsAggsConfig(
-      column,
-      columnId,
-      indexPattern
-    );
-  }
-
   const columnEntries = columnOrder.map((colId) => [colId, columns[colId]] as const);
 
   if (columnEntries.length) {
-    const aggs = columnEntries.map(([colId, col]) => {
-      return getEsAggsConfig(col, colId);
+    const aggs: unknown[] = [];
+    const expressions: ExpressionFunctionAST[] = [];
+    columnEntries.forEach(([colId, col]) => {
+      const def = operationDefinitionMap[col.operationType];
+      if (def.input === 'fullReference') {
+        expressions.push(...def.toExpression(layer, colId, indexPattern));
```

Comment:
Can we be sure the columns are in order so all calculations `col` depends on will be earlier in the list than `col` itself? Otherwise it can happen an expression function tries to act on a column with doesn't exist yet.

We either have to make sure column order is always right or we add a check here and walk the graph of references in direction of the pointers. It's not relevant for single level calculations, but I guess we will run into this in the future.

It could probably be part of the `getColumnOrder` as well

Reply:
Fixed by adding reference sorting to `getColumnOrder`

Reply:
Fixed by updating the getColumnOrder function and calling it in a few more places.